> 패스트캠퍼스 바이트디그리 데이터사이언스 과정 Python Pandas 수업 내용을 정리한 자료입니다.

### transform 함수의 이해 및 활용
#### `transform` 함수
  - groupby 수행 후, transform 함수를 사용하면 원래 index를 유지한 상태로 집계함수 적용 가능
  - 전체 데이터의 집계가 아닌 각 그룹의 집계 계산
  - 새로 생성된 데이터를 원본 Dataframe과 합치기 쉬움

In [35]:
import math
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [8]:
dataframe = pd.read_csv('./train.csv')
dataframe

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [4]:
dataframe.groupby('Pclass').mean()

PassengerId  Survived        Age     SibSp     Parch       Fare
Pclass                                                                 
1        461.597222  0.629630  38.233441  0.416667  0.356481  84.154687
2        445.956522  0.472826  29.877630  0.402174  0.380435  20.662183
3        439.154786  0.242363  25.140620  0.615071  0.393075  13.675550

#### 숫자형 데이터(Numerical Type)
  - 연속성을 띄는 숫자로 이루어진 데이터
    - ex) Age, Fare 등

In [7]:
train_data['Age']

0      22.0
1      38.0
2      26.0
3      35.0
4      35.0
       ... 
886    27.0
887    19.0
888     NaN
889    26.0
890    32.0
Name: Age, Length: 891, dtype: float64

#### 범주형 데이터(Categorical Type)
  - 연속적이지 않은 값(대부분의 경우 숫자형 데이터를 제외한 값)
  - 숫자형 데이터라도 특정 범주를 나타내는 경우 범주형 데이터에 속함
    - ex) Pclass -> 1, 2, 3 숫자형 데이터로 1등급~3등급의 범주를 나타내기 때문에 범주형 데이터에 속함

In [8]:
train_data['Pclass']

0      3
1      1
2      3
3      1
4      3
      ..
886    2
887    1
888    3
889    1
890    3
Name: Pclass, Length: 891, dtype: int64

---

### 숫자형 데이터를 범주형 데이터로 변환하기

#### Pclass 변환하기
  - astype 함수를 이용하여 타입 변환

In [15]:
train_data

PassengerId  Survived Pclass  \
0              1         0      3   
1              2         1      1   
2              3         1      3   
3              4         1      1   
4              5         0      3   
..           ...       ...    ...   
886          887         0      2   
887          888         1      1   
888          889         0      3   
889          890         1      1   
890          891         0      3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [9]:
train_data.info() # Pclass 데이터 타입: int64

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [11]:
train_data['Pclass'] = train_data['Pclass'].astype(str) # Pclass 데이터 타입을 string으로 변환

In [14]:
train_data.info() # Pclass의 데이터 타입이 기존 int64 -> string으로 변환된 것을 확인할 수 있음

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    object 
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(4), object(6)
memory usage: 83.7+ KB


#### Age 숫자형 데이터를 범주형 데이터로 변환하기
  - 변환 logic을 함수로 작성
  - `apply` 함수를 이용하여 적용

In [27]:
def age_int_to_categorical(age):
    if math.isnan(age): # NaN 처리
        return -1
    return math.floor(age / 10) * 10

In [23]:
train_data

PassengerId  Survived Pclass  \
0              1         0      3   
1              2         1      1   
2              3         1      3   
3              4         1      1   
4              5         0      3   
..           ...       ...    ...   
886          887         0      2   
887          888         1      1   
888          889         0      3   
889          890         1      1   
890          891         0      3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [24]:
train_data['Age'].apply(age_int_to_categorical)

0      20
1      30
2      20
3      30
4      30
       ..
886    20
887    10
888    -1
889    20
890    30
Name: Age, Length: 891, dtype: int64

---

### 범주형 데이터를 숫자형 데이터로 변환하기
  - `One-Hot Encoding`

In [28]:
import pandas as pd

In [30]:
train_data = pd.read_csv('./train.csv')
train_data.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

#### One-Hot Encoding
  - 데이터가 특정 범주에 속할 경우 1, 속하지 않을 경우 0으로 표현하는 방법
  - `pandas.get_dummies` 함수를 이용
    - `drop_first`: 첫 번째 카테고리는 사용하지 않음

In [34]:
pd.get_dummies(train_data['Pclass']) # 각 row가 어떤 Pclass에 속하는지 One-Hot Encoding으로 표현, 범주(1, 2, 3)

1  2  3
0    0  0  1
1    1  0  0
2    0  0  1
3    1  0  0
4    0  0  1
..  .. .. ..
886  0  1  0
887  1  0  0
888  0  0  1
889  1  0  0
890  0  0  1

[891 rows x 3 columns]

In [35]:
pd.get_dummies(train_data, columns=['Pclass'])

PassengerId  Survived                                               Name  \
0              1         0                            Braund, Mr. Owen Harris   
1              2         1  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2              3         1                             Heikkinen, Miss. Laina   
3              4         1       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4              5         0                           Allen, Mr. William Henry   
..           ...       ...                                                ...   
886          887         0                              Montvila, Rev. Juozas   
887          888         1                       Graham, Miss. Margaret Edith   
888          889         0           Johnston, Miss. Catherine Helen "Carrie"   
889          890         1                              Behr, Mr. Karl Howell   
890          891         0                                Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch            Ticket     Fare Cabin Embarked  \
0      male  22.0      1      0         A/5 21171   7.2500   NaN        S   
1    female  38.0      1      0          PC 17599  71.2833   C85        C   
2    female  26.0      0      0  STON/O2. 3101282   7.9250   NaN        S   
3    female  35.0      1      0            113803  53.1000  C123        S   
4      male  35.0      0      0            373450   8.0500   NaN        S   
..      ...   ...    ...    ...               ...      ...   ...      ...   
886    male  27.0      0      0            211536  13.0000   NaN        S   
887  female  19.0      0      0            112053  30.0000   B42        S   
888  female   NaN      1      2        W./C. 6607  23.4500   NaN        S   
889    male  26.0      0      0            111369  30.0000  C148        C   
890    male  32.0      0      0            370376   7.7500   NaN        Q   

     Pclass_1  Pclass_2  Pclass_3  
0           0         0         1  
1           1         0         0  
2           0         0         1  
3           1         0         0  
4           0         0         1  
..        ...       ...       ...  
886         0         1         0  
887         1         0         0  
888         0         0         1  
889         1         0         0  
890         0         0         1  

[891 rows x 14 columns]

In [36]:
pd.get_dummies(train_data, columns=['Pclass'], drop_first=True)

PassengerId  Survived                                               Name  \
0              1         0                            Braund, Mr. Owen Harris   
1              2         1  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2              3         1                             Heikkinen, Miss. Laina   
3              4         1       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4              5         0                           Allen, Mr. William Henry   
..           ...       ...                                                ...   
886          887         0                              Montvila, Rev. Juozas   
887          888         1                       Graham, Miss. Margaret Edith   
888          889         0           Johnston, Miss. Catherine Helen "Carrie"   
889          890         1                              Behr, Mr. Karl Howell   
890          891         0                                Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch            Ticket     Fare Cabin Embarked  \
0      male  22.0      1      0         A/5 21171   7.2500   NaN        S   
1    female  38.0      1      0          PC 17599  71.2833   C85        C   
2    female  26.0      0      0  STON/O2. 3101282   7.9250   NaN        S   
3    female  35.0      1      0            113803  53.1000  C123        S   
4      male  35.0      0      0            373450   8.0500   NaN        S   
..      ...   ...    ...    ...               ...      ...   ...      ...   
886    male  27.0      0      0            211536  13.0000   NaN        S   
887  female  19.0      0      0            112053  30.0000   B42        S   
888  female   NaN      1      2        W./C. 6607  23.4500   NaN        S   
889    male  26.0      0      0            111369  30.0000  C148        C   
890    male  32.0      0      0            370376   7.7500   NaN        Q   

     Pclass_2  Pclass_3  
0           0         1  
1           0         0  
2           0         1  
3           0         0  
4           0         1  
..        ...       ...  
886         1         0  
887         0         0  
888         0         1  
889         0         0  
890         0         1  

[891 rows x 13 columns]

---

### DataFrame `group by`
  - 특정 컬럼을 기준으로 그룹화 
  - SQL group by 절과 동일
  - 데이터 분할/병합, 집계연산

In [38]:
import pandas as pd

data_frame = pd.read_csv('./train.csv')
data_frame.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [39]:
class_group = data_frame.groupby('Pclass')
class_group

#### groupby의 groups 속성
  - 각 그룹과 그룹에 속한 index를 dict 형태로 반환

In [59]:
class_group.groups # Pclass는 1, 2, 3의 그룹이 존재 -> dict의 key 값

{1: Int64Index([  1,   3,   6,  11,  23,  27,  30,  31,  34,  35,
             ...
             853, 856, 857, 862, 867, 871, 872, 879, 887, 889],
            dtype='int64', length=216),
 2: Int64Index([  9,  15,  17,  20,  21,  33,  41,  43,  53,  56,
             ...
             848, 854, 861, 864, 865, 866, 874, 880, 883, 886],
            dtype='int64', length=184),
 3: Int64Index([  0,   2,   4,   5,   7,   8,  10,  12,  13,  14,
             ...
             875, 876, 877, 878, 881, 882, 884, 885, 888, 890],
            dtype='int64', length=491)}

In [60]:
gender_group = data_frame.groupby('Sex')
gender_group.groups

{'female': Int64Index([  1,   2,   3,   8,   9,  10,  11,  14,  15,  18,
             ...
             866, 871, 874, 875, 879, 880, 882, 885, 887, 888],
            dtype='int64', length=314),
 'male': Int64Index([  0,   4,   5,   6,   7,  12,  13,  16,  17,  20,
             ...
             873, 876, 877, 878, 881, 883, 884, 886, 889, 890],
            dtype='int64', length=577)}

#### groupping 함수
  - 그룹 데이터에 적용할 수 있는 집계함수(집계함수 적용 시, NaN 값은 집계에서 제외)
    - sum: 데이터의 합
    - count: 데이터 개수
    - mean, std, var: 평균/표준편차/분산
    - min, max: 최소/최대

In [61]:
class_group.count()

PassengerId  Survived  Name  Sex  Age  SibSp  Parch  Ticket  Fare  \
Pclass                                                                      
1               216       216   216  216  186    216    216     216   216   
2               184       184   184  184  173    184    184     184   184   
3               491       491   491  491  355    491    491     491   491   

        Cabin  Embarked  
Pclass                   
1         176       214  
2          16       184  
3          12       491

In [62]:
class_group.sum()

PassengerId  Survived      Age  SibSp  Parch        Fare
Pclass                                                          
1             99705       136  7111.42     90     77  18177.4125
2             82056        87  5168.83     74     70   3801.8417
3            215625       119  8924.92    302    193   6714.6951

In [63]:
class_group.mean()['Age'] # 좌석등급별 평균

Pclass
1    38.233441
2    29.877630
3    25.140620
Name: Age, dtype: float64

In [64]:
class_group.mean()['Survived'] # 좌석등급별 생존률

Pclass
1    0.629630
2    0.472826
3    0.242363
Name: Survived, dtype: float64

In [71]:
data_frame.groupby('Sex').mean()['Survived'] # 성별 생존률

Sex
female    0.742038
male      0.188908
Name: Survived, dtype: float64

#### 여러개의 Column을 선택하여 그룹화하기
  - groupby에 column을 리스트 형태로 전달
  - 집계함수를 적용한 결과는 multi index를 갖는 dataframe

In [75]:
data_frame.groupby(['Pclass', 'Sex']).mean() # 좌석등급과 성별로 그룹화한 결과의 평균 값을 출력

PassengerId  Survived        Age     SibSp     Parch  \
Pclass Sex                                                            
1      female   469.212766  0.968085  34.611765  0.553191  0.457447   
       male     455.729508  0.368852  41.281386  0.311475  0.278689   
2      female   443.105263  0.921053  28.722973  0.486842  0.605263   
       male     447.962963  0.157407  30.740707  0.342593  0.222222   
3      female   399.729167  0.500000  21.750000  0.895833  0.798611   
       male     455.515850  0.135447  26.507589  0.498559  0.224784   

                     Fare  
Pclass Sex                 
1      female  106.125798  
       male     67.226127  
2      female   21.970121  
       male     19.741782  
3      female   16.118810  
       male     12.661633

In [77]:
data_frame.groupby(['Pclass', 'Sex']).mean().index

MultiIndex([(1, 'female'),
            (1,   'male'),
            (2, 'female'),
            (2,   'male'),
            (3, 'female'),
            (3,   'male')],
           names=['Pclass', 'Sex'])

In [80]:
data_frame.groupby(['Pclass', 'Sex']).mean().loc[(2, 'female')] # 특정 row 검색

PassengerId    443.105263
Survived         0.921053
Age             28.722973
SibSp            0.486842
Parch            0.605263
Fare            21.970121
Name: (2, female), dtype: float64

---

#### index를 이용한 groupby
  - index가 있는 경우, groupby에 level 사용 가능
    - level은 index의 depth를 의미
  - `set_index` 함수
    - Column 데이터를 index 레벨로 변경
  - `reset_index` 함수
    - index 초기화

In [81]:
data_frame.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [83]:
data_frame.set_index('Pclass') # Pclass를 index로 이용

PassengerId  Survived  \
Pclass                          
3                 1         0   
1                 2         1   
3                 3         1   
1                 4         1   
3                 5         0   
...             ...       ...   
2               887         0   
1               888         1   
3               889         0   
1               890         1   
3               891         0   

                                                     Name     Sex   Age  \
Pclass                                                                    
3                                 Braund, Mr. Owen Harris    male  22.0   
1       Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                  Heikkinen, Miss. Laina  female  26.0   
1            Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
3                                Allen, Mr. William Henry    male  35.0   
...                                                   ...     ...   ...   
2                                   Montvila, Rev. Juozas    male  27.0   
1                            Graham, Miss. Margaret Edith  female  19.0   
3                Johnston, Miss. Catherine Helen "Carrie"  female   NaN   
1                                   Behr, Mr. Karl Howell    male  26.0   
3                                     Dooley, Mr. Patrick    male  32.0   

        SibSp  Parch            Ticket     Fare Cabin Embarked  
Pclass                                                          
3           1      0         A/5 21171   7.2500   NaN        S  
1           1      0          PC 17599  71.2833   C85        C  
3           0      0  STON/O2. 3101282   7.9250   NaN        S  
1           1      0            113803  53.1000  C123        S  
3           0      0            373450   8.0500   NaN        S  
...       ...    ...               ...      ...   ...      ...  
2           0      0            211536  13.0000   NaN        S  
1           0      0            112053  30.0000   B42        S  
3           1      2        W./C. 6607  23.4500   NaN        S  
1           0      0            111369  30.0000  C148        C  
3           0      0            370376   7.7500   NaN        Q  

[891 rows x 11 columns]

In [84]:
data_frame.set_index(['Pclass', 'Sex']) # Multi Index

PassengerId  Survived  \
Pclass Sex                             
3      male              1         0   
1      female            2         1   
3      female            3         1   
1      female            4         1   
3      male              5         0   
...                    ...       ...   
2      male            887         0   
1      female          888         1   
3      female          889         0   
1      male            890         1   
3      male            891         0   

                                                            Name   Age  SibSp  \
Pclass Sex                                                                      
3      male                              Braund, Mr. Owen Harris  22.0      1   
1      female  Cumings, Mrs. John Bradley (Florence Briggs Th...  38.0      1   
3      female                             Heikkinen, Miss. Laina  26.0      0   
1      female       Futrelle, Mrs. Jacques Heath (Lily May Peel)  35.0      1   
3      male                             Allen, Mr. William Henry  35.0      0   
...                                                          ...   ...    ...   
2      male                                Montvila, Rev. Juozas  27.0      0   
1      female                       Graham, Miss. Margaret Edith  19.0      0   
3      female           Johnston, Miss. Catherine Helen "Carrie"   NaN      1   
1      male                                Behr, Mr. Karl Howell  26.0      0   
3      male                                  Dooley, Mr. Patrick  32.0      0   

               Parch            Ticket     Fare Cabin Embarked  
Pclass Sex                                                      
3      male        0         A/5 21171   7.2500   NaN        S  
1      female      0          PC 17599  71.2833   C85        C  
3      female      0  STON/O2. 3101282   7.9250   NaN        S  
1      female      0            113803  53.1000  C123        S  
3      male        0            373450   8.0500   NaN        S  
...              ...               ...      ...   ...      ...  
2      male        0            211536  13.0000   NaN        S  
1      female      0            112053  30.0000   B42        S  
3      female      2        W./C. 6607  23.4500   NaN        S  
1      male        0            111369  30.0000  C148        C  
3      male        0            370376   7.7500   NaN        Q  

[891 rows x 10 columns]

In [85]:
data_frame.set_index(['Pclass', 'Sex']).reset_index() # index 초기화 

Pclass     Sex  PassengerId  Survived  \
0         3    male            1         0   
1         1  female            2         1   
2         3  female            3         1   
3         1  female            4         1   
4         3    male            5         0   
..      ...     ...          ...       ...   
886       2    male          887         0   
887       1  female          888         1   
888       3  female          889         0   
889       1    male          890         1   
890       3    male          891         0   

                                                  Name   Age  SibSp  Parch  \
0                              Braund, Mr. Owen Harris  22.0      1      0   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  38.0      1      0   
2                               Heikkinen, Miss. Laina  26.0      0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  35.0      1      0   
4                             Allen, Mr. William Henry  35.0      0      0   
..                                                 ...   ...    ...    ...   
886                              Montvila, Rev. Juozas  27.0      0      0   
887                       Graham, Miss. Margaret Edith  19.0      0      0   
888           Johnston, Miss. Catherine Helen "Carrie"   NaN      1      2   
889                              Behr, Mr. Karl Howell  26.0      0      0   
890                                Dooley, Mr. Patrick  32.0      0      0   

               Ticket     Fare Cabin Embarked  
0           A/5 21171   7.2500   NaN        S  
1            PC 17599  71.2833   C85        C  
2    STON/O2. 3101282   7.9250   NaN        S  
3              113803  53.1000  C123        S  
4              373450   8.0500   NaN        S  
..                ...      ...   ...      ...  
886            211536  13.0000   NaN        S  
887            112053  30.0000   B42        S  
888        W./C. 6607  23.4500   NaN        S  
889            111369  30.0000  C148        C  
890            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [87]:
data_frame.set_index('Age').groupby(level=0).mean() # Age가 첫 번째 index가 되고 그룹화를 수행한 뒤, 평균 값을 구함

PassengerId  Survived  Pclass  SibSp  Parch      Fare
Age                                                         
0.42         804.0       1.0     3.0    0.0    1.0    8.5167
0.67         756.0       1.0     2.0    1.0    1.0   14.5000
0.75         557.5       1.0     3.0    2.0    1.0   19.2583
0.83         455.5       1.0     2.0    0.5    1.5   23.8750
0.92         306.0       1.0     1.0    1.0    2.0  151.5500
...            ...       ...     ...    ...    ...       ...
70.00        709.5       0.0     1.5    0.5    0.5   40.7500
70.50        117.0       0.0     3.0    0.0    0.0    7.7500
71.00        295.5       0.0     1.0    0.0    0.0   42.0792
74.00        852.0       0.0     3.0    0.0    0.0    7.7750
80.00        631.0       1.0     1.0    0.0    0.0   30.0000

[88 rows x 6 columns]

In [92]:
def age_to_categorize(age):
    if math.isnan(age):
        return -1
    return math.floor(age / 10) * 10 # 연령대 구하는 공식

In [93]:
# 나이대별 생존률 구하기
data_frame.set_index('Age').groupby(age_to_categorize).mean()

PassengerId  Survived    Pclass     SibSp     Parch       Fare
-1    435.581921  0.293785  2.598870  0.564972  0.180791  22.158567
 0    424.741935  0.612903  2.629032  1.854839  1.403226  30.576679
 10   444.362745  0.401961  2.470588  0.666667  0.470588  32.535132
 20   433.231818  0.350000  2.450000  0.322727  0.250000  27.278937
 30   472.449102  0.437126  2.113772  0.353293  0.329341  40.377294
 40   465.606742  0.382022  1.966292  0.370787  0.471910  38.002297
 50   440.187500  0.416667  1.562500  0.291667  0.270833  47.933333
 60   433.736842  0.315789  1.473684  0.263158  0.368421  48.367542
 70   496.500000  0.000000  1.833333  0.166667  0.166667  30.197233
 80   631.000000  1.000000  1.000000  0.000000  0.000000  30.000000

#### Multiindex를 이용한 grouping

In [94]:
data_frame.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [97]:
data_frame.set_index(['Pclass', 'Sex']).groupby(level=[0, 1]).mean()

PassengerId  Survived        Age     SibSp     Parch  \
Pclass Sex                                                            
1      female   469.212766  0.968085  34.611765  0.553191  0.457447   
       male     455.729508  0.368852  41.281386  0.311475  0.278689   
2      female   443.105263  0.921053  28.722973  0.486842  0.605263   
       male     447.962963  0.157407  30.740707  0.342593  0.222222   
3      female   399.729167  0.500000  21.750000  0.895833  0.798611   
       male     455.515850  0.135447  26.507589  0.498559  0.224784   

                     Fare  
Pclass Sex                 
1      female  106.125798  
       male     67.226127  
2      female   21.970121  
       male     19.741782  
3      female   16.118810  
       male     12.661633

#### 집계(Aggregate) 함수

In [102]:
data_frame.set_index(['Pclass', 'Sex']).groupby(level=[0, 1]).aggregate([np.mean, np.sum, np.max]) # numpy를 이용한 집계함수 활용 가능

PassengerId               Survived                 Age           \
                     mean     sum amax      mean sum amax       mean      sum   
Pclass Sex                                                                      
1      female  469.212766   44106  888  0.968085  91    1  34.611765  2942.00   
       male    455.729508   55599  890  0.368852  45    1  41.281386  4169.42   
2      female  443.105263   33676  881  0.921053  70    1  28.722973  2125.50   
       male    447.962963   48380  887  0.157407  17    1  30.740707  3043.33   
3      female  399.729167   57561  889  0.500000  72    1  21.750000  2218.50   
       male    455.515850  158064  891  0.135447  47    1  26.507589  6706.42   

                        SibSp               Parch                  Fare  \
               amax      mean  sum amax      mean  sum amax        mean   
Pclass Sex                                                                
1      female  63.0  0.553191   52    3  0.457447   43    2  106.125798   
       male    80.0  0.311475   38    3  0.278689   34    4   67.226127   
2      female  57.0  0.486842   37    3  0.605263   46    3   21.970121   
       male    70.0  0.342593   37    2  0.222222   24    2   19.741782   
3      female  63.0  0.895833  129    8  0.798611  115    6   16.118810   
       male    74.0  0.498559  173    8  0.224784   78    5   12.661633   

                                    
                     sum      amax  
Pclass Sex                          
1      female  9975.8250  512.3292  
       male    8201.5875  512.3292  
2      female  1669.7292   65.0000  
       male    2132.1125   73.5000  
3      female  2321.1086   69.5500  
       male    4393.5865   69.5500

In [5]:
# transform 함수 이용
dataframe.groupby('Pclass').transform(np.mean)

PassengerId  Survived        Age     SibSp     Parch       Fare
0     439.154786  0.242363  25.140620  0.615071  0.393075  13.675550
1     461.597222  0.629630  38.233441  0.416667  0.356481  84.154687
2     439.154786  0.242363  25.140620  0.615071  0.393075  13.675550
3     461.597222  0.629630  38.233441  0.416667  0.356481  84.154687
4     439.154786  0.242363  25.140620  0.615071  0.393075  13.675550
..           ...       ...        ...       ...       ...        ...
886   445.956522  0.472826  29.877630  0.402174  0.380435  20.662183
887   461.597222  0.629630  38.233441  0.416667  0.356481  84.154687
888   439.154786  0.242363  25.140620  0.615071  0.393075  13.675550
889   461.597222  0.629630  38.233441  0.416667  0.356481  84.154687
890   439.154786  0.242363  25.140620  0.615071  0.393075  13.675550

[891 rows x 6 columns]

`transform` 함수를 이용할 경우 `데이터 전체에 대한 그룹화를 한 결과를 보여주는 것이 아니라 그룹화하여 계산한 값`을 각 행의 값으로 각각 적용되어 수정된 것을 볼 수 있다. 위의 실습에서는 Pclass 1, 2, 3의 전체 평균을 구한 뒤, 해당하는 그룹의 평균 값으로 각 row 값이 변경된 것을 볼 수 있다. 
  - Pclass 1: 461.597222	0.629630	38.233441	0.416667	0.356481	84.154687
  - Pclass 2: 445.956522	0.472826	29.877630	0.402174	0.380435	20.662183
  - Pclass 3: 439.154786	0.242363	25.140620	0.615071	0.393075	13.675550

In [9]:
dataframe['Age2'] = dataframe.groupby('Pclass').transform(np.mean)['Age']

In [11]:
# 마지막 컬럼으로 Age2가 추가된 것을 볼 수 있음 -> Pclass 각 그룹별 평균 나이 값
dataframe

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked       Age2  
0        0         A/5 21171   7.2500   NaN        S  25.140620  
1        0          PC 17599  71.2833   C85        C  38.233441  
2        0  STON/O2. 3101282   7.9250   NaN        S  25.140620  
3        0            113803  53.1000  C123        S  38.233441  
4        0            373450   8.0500   NaN        S  25.140620  
..     ...               ...      ...   ...      ...        ...  
886      0            211536  13.0000   NaN        S  29.877630  
887      0            112053  30.0000   B42        S  38.233441  
888      2        W./C. 6607  23.4500   NaN        S  25.140620  
889      0            111369  30.0000  C148        C  38.233441  
890      0            370376   7.7500   NaN        Q  25.140620  

[891 rows x 13 columns]

---

### pivot, pivot_table 함수

In [12]:
import numpy as np
import pandas as pd

In [21]:
df = pd.DataFrame({
    '지역': ['서울', '서울', '서울', '경기', '경기', '부산', '서울', '서울', '부산', '경기', '경기', '경기'],
    '요일': ['월요일', '화요일', '수요일', '월요일', '화요일', '월요일', '목요일', '금요일', '화요일', '수요일', '목요일', '금요일'],
    '강수량': [100, 80, 1000, 200, 200, 100, 50, 100, 200, 100, 50, 100],
    '강수확률': [80, 70, 90, 10, 20, 30, 50, 90, 20, 80, 50, 10]
})

In [14]:
df

지역   요일   강수량  강수확률
0   서울  월요일   100    80
1   서울  화요일    80    70
2   서울  수요일  1000    90
3   경기  월요일   200    10
4   경기  화요일   200    20
5   부산  월요일   100    30
6   서울  목요일    50    50
7   서울  금요일   100    90
8   부산  화요일   200    20
9   경기  수요일   100    80
10  경기  목요일    50    50
11  경기  금요일   100    10

#### pivot
  - dataframe의 형태를 변경
  - index, column, 데이터로 사용할 컬럼 명시

In [16]:
df.pivot('지역', '요일') # 첫 번째 parameter는 index 지정, 두 번째 parameter는 column 지정

강수량                              강수확률                        
요일    금요일   목요일     수요일    월요일    화요일   금요일   목요일   수요일   월요일   화요일
지역                                                                 
경기  100.0  50.0   100.0  200.0  200.0  10.0  50.0  80.0  10.0  20.0
부산    NaN   NaN     NaN  100.0  200.0   NaN   NaN   NaN  30.0  20.0
서울  100.0  50.0  1000.0  100.0   80.0  90.0  50.0  90.0  80.0  70.0

In [17]:
df.pivot('요일', '지역')

강수량                 강수확률            
지역      경기     부산      서울    경기    부산    서울
요일                                         
금요일  100.0    NaN   100.0  10.0   NaN  90.0
목요일   50.0    NaN    50.0  50.0   NaN  50.0
수요일  100.0    NaN  1000.0  80.0   NaN  90.0
월요일  200.0  100.0   100.0  10.0  30.0  80.0
화요일  200.0  200.0    80.0  20.0  20.0  70.0

In [18]:
df.pivot('요일', '지역', '강수량')

지역      경기     부산      서울
요일                       
금요일  100.0    NaN   100.0
목요일   50.0    NaN    50.0
수요일  100.0    NaN  1000.0
월요일  200.0  100.0   100.0
화요일  200.0  200.0    80.0

#### pivot_table
  - pivot 함수와 기능적으로 동일
  - 중복데이터에 대한 처리 가능
    - 중복되는 값이 있는 경우, aggregation 함수를 사용하여 값을 채워 넣을 수 있음

In [22]:
df = pd.DataFrame({
    '지역': ['서울', '서울', '서울', '경기', '경기', '부산', '서울', '서울', '부산', '경기', '경기', '경기'],
    '요일': ['월요일', '월요일', '수요일', '월요일', '화요일', '월요일', '목요일', '금요일', '화요일', '수요일', '목요일', '금요일'],
    '강수량': [100, 80, 1000, 200, 200, 100, 50, 100, 200, 100, 50, 100],
    '강수확률': [80, 70, 90, 10, 20, 30, 50, 90, 20, 80, 50, 10]
})

In [23]:
# 중복허용 후 호출
df.pivot_table('요일', '지역') # 서울, 월요일 값이 2개로 중복

DataError: No numeric types to aggregate

In [25]:
# 서울, 월요일이 2개로 중복이었는데 pivot_table을 이용할 경우 aggregation 함수를 이용하여 평균 값 등을 임의로 넣어줄 수 있다.
pd.pivot_table(df, index='요일', columns='지역', aggfunc=np.mean)

강수량                 강수확률            
지역      경기     부산      서울    경기    부산    서울
요일                                         
금요일  100.0    NaN   100.0  10.0   NaN  90.0
목요일   50.0    NaN    50.0  50.0   NaN  50.0
수요일  100.0    NaN  1000.0  80.0   NaN  90.0
월요일  200.0  100.0    90.0  10.0  30.0  75.0
화요일  200.0  200.0     NaN  20.0  20.0   NaN

---

### stack, unstack 함수
#### stack & unstack
  - stack: 컬럼 레벨에서 인덱스 레벨로 dataframe 구조 변경
  - unstack: 인덱스 레벨에서 컬럼 레벨로 dataframe 구조 변경

In [28]:
import numpy as np
import pandas as pd

# sample dataframe
df = pd.DataFrame({
    '지역': ['서울', '서울', '서울', '경기', '경기', '부산', '서울', '서울', '부산', '경기', '경기', '경기'],
    '요일': ['월요일', '화요일', '수요일', '월요일', '화요일', '월요일', '목요일', '금요일', '화요일', '수요일', '목요일', '금요일'],
    '강수량': [100, 80, 1000, 200, 200, 100, 50, 100, 200, 100, 50, 100],
    '강수확률': [80, 70, 90, 10, 20, 30, 50, 90, 20, 80, 50, 10]
})

In [27]:
df

지역   요일   강수량  강수확률
0   서울  월요일   100    80
1   서울  화요일    80    70
2   서울  수요일  1000    90
3   경기  월요일   200    10
4   경기  화요일   200    20
5   부산  월요일   100    30
6   서울  목요일    50    50
7   서울  금요일   100    90
8   부산  화요일   200    20
9   경기  수요일   100    80
10  경기  목요일    50    50
11  경기  금요일   100    10

In [30]:
new_df = df.set_index(['지역', '요일']) # multi-index
new_df

강수량  강수확률
지역 요일             
서울 월요일   100    80
   화요일    80    70
   수요일  1000    90
경기 월요일   200    10
   화요일   200    20
부산 월요일   100    30
서울 목요일    50    50
   금요일   100    90
부산 화요일   200    20
경기 수요일   100    80
   목요일    50    50
   금요일   100    10

In [31]:
new_df.unstack(0)

강수량                 강수확률            
지역      경기     부산      서울    경기    부산    서울
요일                                         
금요일  100.0    NaN   100.0  10.0   NaN  90.0
목요일   50.0    NaN    50.0  50.0   NaN  50.0
수요일  100.0    NaN  1000.0  80.0   NaN  90.0
월요일  200.0  100.0   100.0  10.0  30.0  80.0
화요일  200.0  200.0    80.0  20.0  20.0  70.0

In [32]:
new_df.unstack(1)

강수량                              강수확률                        
요일    금요일   목요일     수요일    월요일    화요일   금요일   목요일   수요일   월요일   화요일
지역                                                                 
경기  100.0  50.0   100.0  200.0  200.0  10.0  50.0  80.0  10.0  20.0
부산    NaN   NaN     NaN  100.0  200.0   NaN   NaN   NaN  30.0  20.0
서울  100.0  50.0  1000.0  100.0   80.0  90.0  50.0  90.0  80.0  70.0

In [33]:
new_df.unstack(0).stack(0)

지역           경기     부산      서울
요일                            
금요일 강수량   100.0    NaN   100.0
    강수확률   10.0    NaN    90.0
목요일 강수량    50.0    NaN    50.0
    강수확률   50.0    NaN    50.0
수요일 강수량   100.0    NaN  1000.0
    강수확률   80.0    NaN    90.0
월요일 강수량   200.0  100.0   100.0
    강수확률   10.0   30.0    80.0
화요일 강수량   200.0  200.0    80.0
    강수확률   20.0   20.0    70.0

In [34]:
new_df.unstack(0).stack(1)

강수량  강수확률
요일  지역              
금요일 경기   100.0  10.0
    서울   100.0  90.0
목요일 경기    50.0  50.0
    서울    50.0  50.0
수요일 경기   100.0  80.0
    서울  1000.0  90.0
월요일 경기   200.0  10.0
    부산   100.0  30.0
    서울   100.0  80.0
화요일 경기   200.0  20.0
    부산   200.0  20.0
    서울    80.0  70.0

### Concat 함수를 이용한 DataFrame 병합
  - SQL에서의 Join과 유사한 기능
    - 단순히 2개 이상의 DataFrame을 합쳐주는 기능
    - 기준 없이 그냥 합치는 기능
  - `pandas.concat` 함수
    - axis 지정에 따라 행 또는 열 기준으로 병합 가능(axis=0 : 행단위 병합)

#### column명이 동일한 2개의 DataFrame 생성

In [37]:
df1 = pd.DataFrame({'key1': np.arange(10), 'value1': np.random.randn(10)})

In [38]:
df2 = pd.DataFrame({'key1': np.arange(10), 'value1': np.random.randn(10)})

In [39]:
df1, df2

(   key1    value1
 0     0  0.487510
 1     1  0.792960
 2     2 -0.826497
 3     3  0.587356
 4     4  2.653618
 5     5 -1.758608
 6     6 -1.436478
 7     7 -0.580494
 8     8 -0.079286
 9     9  0.193913,
    key1    value1
 0     0 -0.940960
 1     1  1.377880
 2     2  0.289555
 3     3 -1.236371
 4     4 -0.740283
 5     5 -0.659854
 6     6  0.941814
 7     7  1.855351
 8     8 -1.372530
 9     9 -0.305672)

In [42]:
pd.concat([df1, df2]) # 두 데이터를 행을 기준으로 병합

key1    value1
0     0  0.487510
1     1  0.792960
2     2 -0.826497
3     3  0.587356
4     4  2.653618
5     5 -1.758608
6     6 -1.436478
7     7 -0.580494
8     8 -0.079286
9     9  0.193913
0     0 -0.940960
1     1  1.377880
2     2  0.289555
3     3 -1.236371
4     4 -0.740283
5     5 -0.659854
6     6  0.941814
7     7  1.855351
8     8 -1.372530
9     9 -0.305672

In [43]:
# ignore_index=True로 지정할 경우 index가 새롭게 생성된다.
pd.concat([df1, df2], ignore_index=True)

key1    value1
0      0  0.487510
1      1  0.792960
2      2 -0.826497
3      3  0.587356
4      4  2.653618
5      5 -1.758608
6      6 -1.436478
7      7 -0.580494
8      8 -0.079286
9      9  0.193913
10     0 -0.940960
11     1  1.377880
12     2  0.289555
13     3 -1.236371
14     4 -0.740283
15     5 -0.659854
16     6  0.941814
17     7  1.855351
18     8 -1.372530
19     9 -0.305672

In [44]:
# 열을 기준으로 병합
pd.concat([df1, df2], axis=1)

key1    value1  key1    value1
0     0  0.487510     0 -0.940960
1     1  0.792960     1  1.377880
2     2 -0.826497     2  0.289555
3     3  0.587356     3 -1.236371
4     4  2.653618     4 -0.740283
5     5 -1.758608     5 -0.659854
6     6 -1.436478     6  0.941814
7     7 -0.580494     7  1.855351
8     8 -0.079286     8 -1.372530
9     9  0.193913     9 -0.305672

#### column명이 다른 경우

In [47]:
df1 = pd.DataFrame({'key1': np.arange(10), 'value1': np.random.randn(10)})
df3 = pd.DataFrame({'key2': np.arange(10), 'value2': np.random.randn(10)})

In [51]:
pd.concat([df1, df3])

key1    value1  key2    value2
0   0.0 -0.983635   NaN       NaN
1   1.0 -0.150554   NaN       NaN
2   2.0  0.846284   NaN       NaN
3   3.0  0.750364   NaN       NaN
4   4.0  0.338422   NaN       NaN
5   5.0 -0.145587   NaN       NaN
6   6.0  1.195911   NaN       NaN
7   7.0  1.062209   NaN       NaN
8   8.0 -0.124973   NaN       NaN
9   9.0  0.479502   NaN       NaN
0   NaN       NaN   0.0 -0.176942
1   NaN       NaN   1.0 -0.247221
2   NaN       NaN   2.0 -0.707694
3   NaN       NaN   3.0 -2.197850
4   NaN       NaN   4.0  0.249342
5   NaN       NaN   5.0 -0.211634
6   NaN       NaN   6.0 -0.472070
7   NaN       NaN   7.0  1.090267
8   NaN       NaN   8.0  1.305794
9   NaN       NaN   9.0  0.787349

In [52]:
pd.concat([df1, df3], axis=1)

key1    value1  key2    value2
0     0 -0.983635     0 -0.176942
1     1 -0.150554     1 -0.247221
2     2  0.846284     2 -0.707694
3     3  0.750364     3 -2.197850
4     4  0.338422     4  0.249342
5     5 -0.145587     5 -0.211634
6     6  1.195911     6 -0.472070
7     7  1.062209     7  1.090267
8     8 -0.124973     8  1.305794
9     9  0.479502     9  0.787349

---

### Merge/Join 함수를 이용한 DataFrame 병합하기
#### merge
  - SQL의 Join과 같이 column을 기준으로 병합
    - join 방식: `how` 파라미터를 지정하여 다양한 종류의 join을 할 수 있음
      - inner join: default, 일치하는 값이 있는 경우 join
      - left: left outer join
      - right: right outer join
      - outer: full outer join
  - `pandas.merge` 함수 이용

In [54]:
customer = pd.DataFrame({'customer_id': np.arange(6), 
                         'name': ['철수', '영희', '길동', '영수', '수민', '동건'], 
                         '나이': [40, 20, 21, 30, 33, 18]})
customer

customer_id name  나이
0            0   철수  40
1            1   영희  20
2            2   길동  21
3            3   영수  30
4            4   수민  33
5            5   동건  18

In [63]:
orders = pd.DataFrame({'customer_id': [1, 1, 2, 2, 2, 3, 3, 1, 4, 9],
                      'item': ['치약', '칫솔', '이어폰', '헤드셋', '수건', '생수', '수건', '치약', '생수', '케이스'],
                       'quantity': [1, 2, 1, 1, 3, 2, 2, 3, 2, 1]
                      })
orders

customer_id item  quantity
0            1   치약         1
1            1   칫솔         2
2            2  이어폰         1
3            2  헤드셋         1
4            2   수건         3
5            3   생수         2
6            3   수건         2
7            1   치약         3
8            4   생수         2
9            9  케이스         1

In [59]:
# on -> join 기준이 되는 컬럼 지정, how 파라미터 지정이 없으면 inner join
pd.merge(customer, orders, on='customer_id') 

customer_id name  나이 item  quantity
0            1   영희  20   치약         1
1            1   영희  20   칫솔         2
2            1   영희  20   치약         3
3            2   길동  21  이어폰         1
4            2   길동  21  헤드셋         1
5            2   길동  21   수건         3
6            3   영수  30   생수         2
7            3   영수  30   수건         2
8            4   수민  33   생수         2

In [61]:
# left outer join: 왼쪽 DataFrame(SQL 테이블)은 그대로 남겨둠(일치하는 값이 없는 경우 NaN)
pd.merge(customer, orders, on='customer_id', how='left') 

customer_id name  나이 item  quantity
0             0   철수  40  NaN       NaN
1             1   영희  20   치약       1.0
2             1   영희  20   칫솔       2.0
3             1   영희  20   치약       3.0
4             2   길동  21  이어폰       1.0
5             2   길동  21  헤드셋       1.0
6             2   길동  21   수건       3.0
7             3   영수  30   생수       2.0
8             3   영수  30   수건       2.0
9             4   수민  33   생수       2.0
10            5   동건  18  NaN       NaN

In [62]:
# right outer join: 오른쪽 DataFrame을 남겨둠
pd.merge(customer, orders, on='customer_id', how='right') 

customer_id name    나이 item  quantity
0            1   영희  20.0   치약         1
1            1   영희  20.0   칫솔         2
2            1   영희  20.0   치약         3
3            2   길동  21.0  이어폰         1
4            2   길동  21.0  헤드셋         1
5            2   길동  21.0   수건         3
6            3   영수  30.0   생수         2
7            3   영수  30.0   수건         2
8            4   수민  33.0   생수         2
9            9  NaN   NaN  케이스         1

In [65]:
# outer join: DataFrame을 그대로 병합
pd.merge(customer, orders, on='customer_id', how='outer') 

customer_id name    나이 item  quantity
0             0   철수  40.0  NaN       NaN
1             1   영희  20.0   치약       1.0
2             1   영희  20.0   칫솔       2.0
3             1   영희  20.0   치약       3.0
4             2   길동  21.0  이어폰       1.0
5             2   길동  21.0  헤드셋       1.0
6             2   길동  21.0   수건       3.0
7             3   영수  30.0   생수       2.0
8             3   영수  30.0   수건       2.0
9             4   수민  33.0   생수       2.0
10            5   동건  18.0  NaN       NaN
11            9  NaN   NaN  케이스       1.0

In [67]:
merged_data = pd.merge(customer, orders, on='customer_id')
merged_data

customer_id name  나이 item  quantity
0            1   영희  20   치약         1
1            1   영희  20   칫솔         2
2            1   영희  20   치약         3
3            2   길동  21  이어폰         1
4            2   길동  21  헤드셋         1
5            2   길동  21   수건         3
6            3   영수  30   생수         2
7            3   영수  30   수건         2
8            4   수민  33   생수         2

In [72]:
merged_data.groupby('item').sum().sort_values(by='quantity', ascending=False)

customer_id  나이  quantity
item                           
수건              5  51         5
생수              7  63         4
치약              2  40         4
칫솔              1  20         2
이어폰             2  21         1
헤드셋             2  21         1

In [73]:
merged_data.groupby(['name', 'item']).sum()

customer_id  나이  quantity
name item                           
길동   수건              2  21         3
     이어폰             2  21         1
     헤드셋             2  21         1
수민   생수              4  33         2
영수   생수              3  30         2
     수건              3  30         2
영희   치약              2  40         4
     칫솔              1  20         2

In [78]:
merged_data.groupby(['name', 'item']).sum().loc['영희', 'quantity']

item
치약    4
칫솔    2
Name: quantity, dtype: int64

---

#### join 함수
  - 내부적으로 pandas.merge 함수 사용
  - 기본적으로 index를 사용하여 left join 수행

In [79]:
cust = customer.set_index('customer_id')
order = orders.set_index('customer_id')

In [80]:
cust.join(order, how='inner')

name  나이 item  quantity
customer_id                        
1             영희  20   치약         1
1             영희  20   칫솔         2
1             영희  20   치약         3
2             길동  21  이어폰         1
2             길동  21  헤드셋         1
2             길동  21   수건         3
3             영수  30   생수         2
3             영수  30   수건         2
4             수민  33   생수         2